sdk/core/core-rest-pipeline/documentation/pipeline.md

```diff
@@ -0,0 +1,251 @@
+# Policies
+
+A pipeline policy manipulates a request as it travels through the pipeline. It is conceptually a middleware that is allowed to modify the request before it is made as well as the response when it is received.
+
+It is important to note that policies have a bubbling nature, this means that the **first** policy to touch the request will be the **last** policy to work on the response.
+
+More information about the Middleware pattern can be found [here](https://docs.microsoft.com/aspnet/core/fundamentals/middleware/?tabs=aspnetcore2x&view=aspnetcore-6.0)
+
+```typescript
+export type SendRequest = (request: PipelineRequest) => Promise<PipelineResponse>;
+
+export interface PipelinePolicy {
+  name: string;
+  sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse>;
+}
+```
+
+# Pipelines
+
+A `Pipeline` composes several policies and provides the ability to make a request that will flow through the policies in a specific order determined by their requirements. Which allows manipulating each request before it is sent to the server and to work on the response after receiving it.
+
+```typescript
+export interface AddPipelineOptions {
+  afterPhase?: PipelinePhase;
+  afterPolicies?: string[];
+  beforePolicies?: string[];
+  phase?: PipelinePhase;
+}
+
+export interface Pipeline {
+  addPolicy(policy: PipelinePolicy, options?: AddPipelineOptions): void;
+  clone(): Pipeline;
+  getOrderedPolicies(): PipelinePolicy[];
+  removePolicy(options: { name?: string; phase?: PipelinePhase }): PipelinePolicy[];
+  sendRequest(httpClient: HttpClient, request: PipelineRequest): Promise<PipelineResponse>;
+}
+```
+
+It is important to never rely on policy insert ordering and always express policy execution constraints using `AddPipelineOptions` when policies have interactions. For example, the following code snippets have identical effects:
+
+```typescript
+pipeline.addPolicy(barPolicy);
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+
+// The below is the same as the above.
+// Policies don't have to be part of the pipeline in order to be
+// referenced by before/afterPolicy.
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+pipeline.addPolicy(barPolicy);
+```
+
+Phases are predefined buckets of activity that pre-defined policies inside core will fall into. They are currently a fixed set and not customizable by pipeline consumers.
+
+```typescript
+export type PipelinePhase = "Deserialize" | "Serialize" | "Retry" | "Sign";
+```
+
+The execution order is:
+
+1. **Serialize Phase** - Policies that assamble the request to be sent through HTTP, i.e. transforming the body object to its JSON string representation.
+2. **Policies not in a phase** - Any policies added to the pipeline without an specified phase
+3. **Deserialize Phase** - Policies that transform the raw HTTP response to a friendlier shapes to the clients. For example transforming the body JSON string to a JavaScript object
```

Comment:
expanding on this a little
```suggestion
3. **Deserialize Phase** - Policies that transform the raw HTTP response into friendlier response shapes. In addition to transforming the body JSON string into a JavaScript object, this can also include things like creating native JS `Date` objects from strings and other object forms not natively supported by JSON.
```